File: bindings/dart/test/wallet_test.dart
import 'dart:io';

import 'package:cdk/cdk.dart';
import 'package:test/test.dart';

void main() {
  late Wallet wallet;
  late String dbPath;

  final mintUrl =
      Platform.environment['CDK_DART_TEST_MINT_URL'] ??
      'https://dummy-mint-url-for-local-testing.invalid';
  final runLiveMintTests =
      Platform.environment['CDK_DART_TEST_MINT_URL']?.isNotEmpty ?? false;
  final settlementDelaySeconds =
      int.tryParse(
        Platform.environment['CDK_DART_MINT_SETTLEMENT_DELAY_SECONDS'] ?? '',
      ) ??
      3;

  setUp(() {
    dbPath =
        '${Directory.systemTemp.path}/${DateTime.now().microsecondsSinceEpoch}.sqlite';
    wallet = Wallet.open(
      request: WalletOpenRequest(
        mintUrl: mintUrl,
        unit: SatCurrencyUnit(),
        mnemonic: generateMnemonic(),
        store: SqliteWalletStore(dbPath),
        config: WalletConfig(targetProofCount: null),
      ),
    );
  });

  tearDown(() {
    wallet.dispose();
    try {
      File(dbPath).deleteSync();
    } catch (_) {}
  });

  test('initial balance is zero', () async {
    final balance = await wallet.balance();
    expect(balance.available.value, equals(0));
    expect(balance.pending.value, equals(0));
    expect(balance.reserved.value, equals(0));
  });

  test('in-memory sqlite handles concurrent access', () async {
    final memoryWallet = Wallet.open(
      request: WalletOpenRequest(
        mintUrl: mintUrl,
        unit: SatCurrencyUnit(),
        mnemonic: generateMnemonic(),
        store: SqliteWalletStore(':memory:'),
        config: WalletConfig(targetProofCount: null),
      ),
    );

    try {
      final balances = await Future.wait(
        List.generate(64, (_) => memoryWallet.balance()),
      );
      for (final balance in balances) {
        expect(balance.available.value, equals(0));
      }
    } finally {
      memoryWallet.dispose();
    }
  });

  Wallet configuredWallet(RateLimit? rateLimit) => Wallet.open(
    request: WalletOpenRequest(
      mintUrl: 'https://mint.example.com',
      unit: SatCurrencyUnit(),
      mnemonic: generateMnemonic(),
      store: SqliteWalletStore(':memory:'),
      config: WalletConfig(
        targetProofCount: null,
        rateLimit: rateLimit,
      ),
    ),
  );

  test('construction accepts supported pacing policies', () {
    for (final policy in <RateLimit>[
      DefaultRateLimit(),
      DisabledRateLimit(),
      CustomRateLimit(capacity: 5, refillPerMinute: 30),
    ]) {
      final opened = configuredWallet(policy);
      expect(opened.identity().mintUrl.url, 'https://mint.example.com');
      opened.dispose();
    }
  });

  test('zero rate limit values are rejected', () {
    expect(
      () => configuredWallet(
        CustomRateLimit(capacity: 0, refillPerMinute: 30),
      ),
      throwsA(isA<FfiException>()),
    );
    expect(
      () => configuredWallet(
        CustomRateLimit(capacity: 5, refillPerMinute: 0),
      ),
      throwsA(isA<FfiException>()),
    );
  });

  test(
    'mint flow',
    () async {
      final session = await wallet.requestMinting(
        request: MintRequest(
          method: Bolt11PaymentMethod(),
          amount: Amount(value: 100),
        ),
      );
      try {
        final initial = session.initialState();
        expect(initial.id, isNotEmpty);
        expect(initial.paymentRequest, isNotEmpty);

        // testnut pays quotes automatically.
        await Future.delayed(Duration(seconds: settlementDelaySeconds));
        await session.refresh();
        final claimed = await session.claim();
        expect(claimed.value, equals(100));

        final balance = await wallet.balance();
        expect(balance.available.value, equals(100));
      } finally {
        session.dispose();
      }
    },
    skip: !runLiveMintTests
        ? 'Set CDK_DART_TEST_MINT_URL to run live mint integration tests'
        : false,
  );
}

File: bindings/kotlin/cdk-jvm/src/test/kotlin/org/cashudevkit/WalletTest.kt
package org.cashudevkit

import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.File

class WalletTest {

    private lateinit var wallet: Wallet
    private lateinit var dbFile: File

    @BeforeEach
    fun setUp() {
        dbFile = File.createTempFile("cdk_test_", ".sqlite")
        wallet = Wallet.open(
            WalletOpenRequest(
                mintUrl = "https://testnut.cashudevkit.org",
                unit = CurrencyUnit.Sat,
                mnemonic = generateMnemonic(),
                store = WalletStore.Sqlite(path = dbFile.absolutePath),
                config = WalletConfig(targetProofCount = null),
            ),
        )
    }

    @AfterEach
    fun tearDown() {
        wallet.close()
        dbFile.delete()
    }

    @Test
    fun `initial balance is zero`() = runBlocking {
        val balance = wallet.balance()
        assertEquals(0UL, balance.available.value)
        assertEquals(0UL, balance.pending.value)
        assertEquals(0UL, balance.reserved.value)
    }

    @Test
    fun `in-memory sqlite handles concurrent access`() = runBlocking {
        val memoryWallet = Wallet.open(
            WalletOpenRequest(
                mintUrl = "https://testnut.cashudevkit.org",
                unit = CurrencyUnit.Sat,
                mnemonic = generateMnemonic(),
                store = WalletStore.Sqlite(path = ":memory:"),
                config = WalletConfig(targetProofCount = null),
            ),
        )

        try {
            val balances = coroutineScope {
                (0 until 64).map {
                    async { memoryWallet.balance() }
                }.awaitAll()
            }
            balances.forEach { balance ->
                assertEquals(0UL, balance.available.value)
            }
        } finally {
            memoryWallet.close()
        }
    }

    private fun configuredWallet(rateLimit: RateLimit?): Wallet = Wallet.open(
        WalletOpenRequest(
            mintUrl = "https://mint.example.com",
            unit = CurrencyUnit.Sat,
            mnemonic = generateMnemonic(),
            store = WalletStore.Sqlite(path = ":memory:"),
            config = WalletConfig(targetProofCount = null, rateLimit = rateLimit),
        ),
    )

    @Test
    fun `construction accepts supported pacing policies`() {
        listOf(
            RateLimit.Default,
            RateLimit.Disabled,
            RateLimit.Custom(capacity = 5U, refillPerMinute = 30U),
        ).forEach { policy ->
            configuredWallet(policy).use { opened ->
                assertEquals("https://mint.example.com", opened.identity().mintUrl.url)
            }
        }
    }

    @Test
    fun `zero rate limit values are rejected`() {
        assertThrows(FfiException::class.java) {
            configuredWallet(RateLimit.Custom(capacity = 0U, refillPerMinute = 30U))
        }
        assertThrows(FfiException::class.java) {
            configuredWallet(RateLimit.Custom(capacity = 5U, refillPerMinute = 0U))
        }
    }

    @Test
    fun `mint flow`() = runBlocking {
        val session = wallet.requestMinting(
            MintRequest(
                method = PaymentMethod.Bolt11,
                amount = Amount(value = 100UL),
            ),
        )
        try {
            val initial = session.initialState()
            assertTrue(initial.id.isNotEmpty())
            assertTrue(initial.paymentRequest.isNotEmpty())

            // testnut pays quotes automatically.
            delay(3000)
            session.refresh()
            val claimed = session.claim()
            assertEquals(100UL, claimed.value)

            val balance = wallet.balance()
            assertEquals(100UL, balance.available.value)
        } finally {
            session.close()
        }
    }
}
